internal/controllers/app_controller.go

```diff
@@ -210,8 +207,27 @@ func (r *AppReconciler) reconcile(ctx context.Context, app *ketchv1.App) reconci
 
 	// check for canary deployment
 	if app.Spec.Canary.Active {
+		// ensures that the canary deployment exists
+		if len(app.Spec.Deployments) <= 1 {
+			app.Spec.Canary.Active = false
+			return reconcileResult{
+				status:     v1.ConditionFalse,
+				message:    "no canary deployment found",
+				useTimeout: true,
+			}
+		}
+
 		// retry until all pods for canary deployment comes to running state.
 		if err := checkPodStatus(r.Client, app.Name, app.Spec.Deployments[1].Version); err != nil {
+			// check for rollback
+			app.CheckForRollback()
+			if e := r.Update(ctx, app); err != nil {
+				return reconcileResult{
+					status:  v1.ConditionFalse,
+					message: fmt.Sprintf("failed to perform rollback: %v", e),
```

Comment:
because `CheckForRollback` function implements two different things this error message has to be "failed to perform rollback" or "failed to increment canary failure counter"